src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/sql.js

```diff
@@ -0,0 +1,395 @@
+        super(api)
+        this.setPreprocessorModule('Standard.Visualization.Sql.Visualization')
+        this.setPreprocessorCode(`x -> here.prepare_visualization x`)
+    }
+
+    onDataReceived(data) {
+        while (this.dom.firstChild) {
+            this.dom.removeChild(this.dom.lastChild)
+        }
+
+        let parsedData = data
+        if (typeof data === 'string') {
+            parsedData = JSON.parse(data)
+        }
+
+        let visHtml = visualizationStyle
+        if (parsedData.error !== undefined) {
+            visHtml += parsedData.error
+        } else {
+            const params = parsedData.interpolations.map(param =>
+                renderInterpolationParameter(this.theme, param)
+            )
+
+            let language = 'sql'
+            if (parsedData.dialect == 'postgresql') {
+                language = 'postgresql'
+            }
+
+            const formatted = sqlFormatter.format(parsedData.code, {
+                params: params,
+                language: language,
+            })
+
+            const codeRepresentation = '<pre class="sql">' + formatted + '</pre>'
+            visHtml += codeRepresentation
+        }
+
+        const containers = this.createContainers()
+        const parentContainer = containers[0]
+        const scrollable = containers[1]
+        scrollable.innerHTML = visHtml
+        this.dom.appendChild(parentContainer)
+
+        const tooltip = new Tooltip(parentContainer)
+        const baseMismatches = this.dom.getElementsByClassName('mismatch')
+        const extendedMismatchAreas = this.dom.getElementsByClassName('mismatch-mouse-area')
+        setupMouseInteractionForMismatches(tooltip, baseMismatches)
+        setupMouseInteractionForMismatches(tooltip, extendedMismatchAreas)
+    }
+
+    /**
+     * Creates containers for the visualization.
+     */
+    createContainers() {
+        const parentContainer = document.createElement('div')
+        parentContainer.setAttributeNS(null, 'style', 'position: relative;')
+        const width = this.dom.getAttributeNS(null, 'width')
+        const height = this.dom.getAttributeNS(null, 'height')
+        const scrollable = document.createElement('div')
+        scrollable.setAttributeNS(null, 'id', 'vis-sql-view')
+        scrollable.setAttributeNS(null, 'class', 'scrollable')
+        scrollable.setAttributeNS(null, 'viewBox', '0 0 ' + width + ' ' + height)
+        scrollable.setAttributeNS(null, 'width', '100%')
+        scrollable.setAttributeNS(null, 'height', '100%')
+        const viewStyle = `width: ${width - 5}px;
+             height: ${height - 5}px;
+             overflow: scroll;
+             padding:2.5px;`
+        scrollable.setAttributeNS(null, 'style', viewStyle)
+        parentContainer.appendChild(scrollable)
+        return [parentContainer, scrollable]
+    }
+
+    setSize(size) {
+        this.dom.setAttributeNS(null, 'width', size[0])
+        this.dom.setAttributeNS(null, 'height', size[1])
+    }
```

Comment:
again, why such strangedesign with arrays instead of objects?

Reply:
This is a function that every `Visualization` must override provided from the Rust API, so it is not my design :) It is the same as in most other visualizations. Interestingly though, now I have noticed that the documentation indeed states that this is an object with `width` and `height` but I have checked the Rust code and it seems to pass arrays.

I guess it should be filed as an issue?